### ISSUE TITLE
Potential false positive for let_and_return

### ISSUE THREAD

**Issue by Keats** [rust-lang/rust-clippy#1322]
I have the following macro:

```rust
#[macro_export]
macro_rules! try_get_value {
    ($filter_name:expr, $var_name:expr, $ty:ty, $val:expr) => {{
        let v: $ty = match ::serde_json::value::from_value($val.clone()) {
            Ok(s) => s,
            Err(_) => {
                return Err(::errors::TeraError::FilterIncorrectArgType(
                    $filter_name.to_string(),
                    $var_name.to_string(),
                    $val,
                    stringify!($ty).to_string())
                );
            }
        };
        v
    }};
}
```

And running clippy, it complains that I can directly return the match but doing so fails since it doesn't know the type (`$ty` can `f32`, `String`, etc).
Afaik there is no other way to cast the `s` variable other than that syntax (`as $ty` doesn't work). Am I missing a way to do it or is it a real false positive?

**Comment by oli-obk** [comment]
it's a false positive, but you can actually improve it in this case, by annotating the `from_value` function call

```
#[macro_export]
macro_rules! try_get_value {
    ($filter_name:expr, $var_name:expr, $ty:ty, $val:expr) => {{
        match ::serde_json::value::from_value::<$ty>($val.clone()) {
            Ok(s) => s,
            Err(_) => {
                return Err(::errors::TeraError::FilterIncorrectArgType(
                    $filter_name.to_string(),
                    $var_name.to_string(),
                    $val,
                    stringify!($ty).to_string())
                );
            }
        }
    }};
}
```

**Comment by Keats** [comment]
Embarrassing that I forgot to try that >_>

Thanks!
Do I leave the issue open for the false positive or close it?

**Comment by oli-obk** [comment]
leave it open, we need to improve the message

**Comment by ebroto** [comment]
I think this issue can be closed because `let_and_return` currently does not lint in the presence of type annotations (since [81b35d1](https://github.com/rust-lang/rust-clippy/commit/81b35d197b7e14149cf6cfc02ed4087f273d6915))
https://github.com/rust-lang/rust-clippy/blob/a8957acb1345448ec94586d48bcc8079bd0a63ab/clippy_lints/src/returns.rs#L200-L201
and it also avoids linting inside macros (since #5008)
https://github.com/rust-lang/rust-clippy/blob/a8957acb1345448ec94586d48bcc8079bd0a63ab/clippy_lints/src/returns.rs#L210
